crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -823,133 +822,167 @@ where
+        while let Some(ref block_import_notification) =
+            block_importing_notification_stream.next().await
         {
-            let block_number_to_archive =
-                match block_number.checked_sub(&confirmation_depth_k.into()) {
-                    Some(block_number_to_archive) => block_number_to_archive,
-                    None => {
-                        continue;
-                    }
-                };
-
-            if best_archived_block_number >= block_number_to_archive {
-                // This block was already archived, skip
-                continue;
-            }
-
-            best_archived_block_number = block_number_to_archive;
+            archive_block(
+                &mut archiver,
+                segment_headers_store.clone(),
+                client.clone(),
+                &sync_oracle,
+                telemetry.clone(),
+                archived_segment_notification_sender.clone(),
+                confirmation_depth_k,
+                &mut best_archived_block_number,
+                &mut best_archived_block_hash,
+                block_import_notification.block_number,
+            )
+            .await?;
+        }
 
-            let block = client
-                .block(
-                    client
-                        .hash(block_number_to_archive)?
-                        .expect("Older block by number must always exist"),
-                )?
-                .expect("Older block by number must always exist");
+        Ok(())
+    })
+}
 
-            let parent_block_hash = *block.block.header().parent_hash();
-            let block_hash_to_archive = block.block.hash();
+#[allow(clippy::too_many_arguments)]
+async fn archive_block<Block, Backend, Client, AS, SO>(
+    archiver: &mut Archiver,
+    segment_headers_store: SegmentHeadersStore<AS>,
+    client: Arc<Client>,
+    sync_oracle: &SubspaceSyncOracle<SO>,
+    telemetry: Option<TelemetryHandle>,
+    archived_segment_notification_sender: SubspaceNotificationSender<ArchivedSegmentNotification>,
+    confirmation_depth_k: BlockNumber,
+    best_archived_block_number: &mut NumberFor<Block>,
+    best_archived_block_hash: &mut Block::Hash,
+    block_number: NumberFor<Block>,
+) -> sp_blockchain::Result<bool>
+where
+    Block: BlockT,
+    Backend: BackendT<Block>,
+    Client: ProvideRuntimeApi<Block>
+        + BlockBackend<Block>
+        + HeaderBackend<Block>
+        + LockImportRun<Block, Backend>
+        + Finalizer<Block, Backend>
+        + AuxStore
+        + Send
+        + Sync
+        + 'static,
+    Client::Api: SubspaceApi<Block, FarmerPublicKey> + ObjectsApi<Block>,
+    AS: AuxStore + Send + Sync + 'static,
+    SO: SyncOracle + Send + Sync + 'static,
+{
+    let block_number_to_archive = match block_number.checked_sub(&confirmation_depth_k.into()) {
+        Some(block_number_to_archive) => block_number_to_archive,
+        None => {
+            return Ok(true);
```

Comment:
NIT: in this refactoring PR the boolean returned is always `true` and no documentation why is it even there or what it means. Since this boolean is only meant to be useful later I probably wouldn't introduce it in the first commit to prevent confusion an only introduce in the second where reviewer can see how it is used and why it is needed.